lib/Modem.ts

```diff
@@ -0,0 +1,829 @@
+
+	// command queue
+	private prioQueue: (Command | prTypes.CmdSequence)[] = [];
+	private queue: (Command | prTypes.CmdSequence)[] = [];
+	private _receivedData = '';
+	private _receivedCmdResponse: CommandResponse = [];
+	private _isLocked = false;
+	private _interval: NodeJS.Timer | null = null;
+
+	constructor(targetDevice: string, mode: ModemMode = 'PDU', options: ModemConstructorOptions = {}) {
+		this._mode = mode;
+		this.pinCode = options.pin || null;
+
+		this.options = {
+			autoDeleteOnReceive: options.autoDeleteOnReceive !== undefined ? options.autoDeleteOnReceive : false,
+			enableConcatenation: options.enableConcatenation !== undefined ? options.enableConcatenation : false,
+			customInitCommand: options.customInitCommand !== undefined ? options.customInitCommand : null,
+			autoInitOnOpen: options.autoInitOnOpen !== undefined ? options.autoInitOnOpen : true,
+			cnmiCommand: options.cnmiCommand !== undefined ? options.cnmiCommand : 'AT+CNMI=2,1,0,2,1'
+		};
+
+		const serialPortOptions: SerialPortOptions = Object.assign(
+			// defaults
+			{
+				baudRate: 9600,
+				dataBits: 8,
+				stopBits: 1,
+				highWaterMark: 16384,
+				parity: 'none',
+				rtscts: false,
+				xon: false,
+				xoff: false
+			},
+			// options
+			options.serialPortOptions,
+			// overide options
+			{
+				path: targetDevice,
+				autoOpen: false
+			}
+		);
+
+		this.port = new SerialPort(serialPortOptions);
+		this.port.on('data', (data) => this._dataReceived(data));
+	}
+
+	/*
+	 * getter
+	 */
+
+	get targetDevice() {
+		return this.port.path;
+	}
+
+	get isOpen() {
+		return this.port.isOpen;
+	}
+
+	get mode() {
+		return this._mode;
+	}
+
+	get queueLength() {
+		return this.prioQueue.length + this.queue.length;
+	}
+
+	/*
+	 * command queue
+	 */
+
+	private pushToQue(cmd: Command | prTypes.CmdSequence, prio = false) {
+		if (prio) {
+			this.prioQueue.push(cmd);
+			return;
+		}
+
+		this.queue.push(cmd);
+	}
+
+	private async _execNext() {
+		if (this._isLocked || !this.port.isOpen) {
+			return;
+		}
+
+		const item = this.prioQueue.shift() || this.queue.shift();
+
+		if (!item) {
+			return;
+		}
+
+		this._isLocked = true;
+
+		if (item instanceof Command) {
+			await this._executeCMD(item);
+		} else {
+			for (const cmd of item.cmds) {
+				const result = await this._executeCMD(cmd);
+
+				if (result instanceof Error && item.cancelOnFailure) {
+					if (item.onFailed) {
+						item.onFailed(result);
+					}
+
+					break;
+				}
+			}
+		}
+
+		this._isLocked = false;
+		return item;
+	}
+
+	private async _executeCMD(cmd: Command) {
+		const result = await new Promise((resolve: (result: CommandResponse | Error) => void) => {
+			if (cmd.deprecated) {
+				return resolve(new Error('Command marked as deprecated'));
+			}
+
+			const write = `${cmd.ATCommand}\r`;
+
+			this.events.emitWriteToModem(write);
+			this.port.write(write);
+
+			const timeout = setTimeout(() => resolve(new Error('Command is timeouted!')), cmd.timeout);
```

Comment:
You shoiuld remember if resolve was called already in timeout and then not resolve below ... 

Yes you catched the case that "real response comes before timeout", but you miss the (yes unlikely but possible) case of timeout triggered before real response?

Reply:
You are right. Do you have a solution to this problem? 🤔

Reply:
Is hopefully improved in the next update